pkg/source/source.go

```diff
@@ -68,164 +77,249 @@ var _ Source = &Channel{}
-			cs.doStop()
-			return
+		case <-cs.ctx.Done():
+			return nil
 		case evt, stillOpen := <-cs.Source:
 			if !stillOpen {
-				// if the source channel is closed, we're never gonna get
-				// anything more on it, so stop & bail
-				cs.doStop()
-				return
+				return nil
 			}
-			cs.distribute(evt)
+
+			func() {
+				cs.mu.Lock()
+				defer cs.mu.Unlock()
+
+				for _, dst := range cs.destinations {
+					// We cannot make it under goroutine here, or we'll meet the
+					// race condition of writing message to closed channels.
+					// To avoid blocking, the dest channels are expected to be of
+					// proper buffer size. If we still see it blocked, then
+					// the controller is thought to be in an abnormal state.
+					dst <- evt
+				}
+			}()
 		}
 	}
 }
 
+// Stop is internal and should be called only by the Controller to stop the Source. It should block until the
+// Source has stopped.
+// Stop can be called multiple times, and should be idempotent. If the Source is not running, Stop will prevent
+// the Source from starting in the future without error.
+func (cs *Channel) Stop() error {
+	cs.mu.Lock()
+
+	// Ensure that when we release the lock, we stop an in-process & future calls to Start().
+	cs.canceled = true
+
+	// Try to cancel the context and wait for the group to finish.
+	if cs.cancel != nil {
+		cs.cancel()
+	}
+
+	if cs.group == nil {
+		cs.mu.Unlock()
+		return nil
+	}
+
+	cs.mu.Unlock()
+	return cs.group.Wait()
+}
+
 // Informer is used to provide a source of events originating inside the cluster from Watches (e.g. Pod Create).
 type Informer struct {
 	// Informer is the controller-runtime Informer
 	Informer cache.Informer
+
+	mu sync.Mutex
+	// canceled is true if the source has been canceled and causes any following/ in-progress
+	// Start calls to no-op.
+	canceled bool
+	// ctx is used as context for dynamically added event handlers.
+	ctx context.Context
+	// cancel is used to cancel the goroutine started by Start.
+	cancel context.CancelFunc
+	// registrations is a list of all registered EventHandler handles.
+	registrations []toolscache.ResourceEventHandlerRegistration
 }
 
 var _ Source = &Informer{}
 
 // Start is internal and should be called only by the Controller to register an EventHandler with the Informer
 // to enqueue reconcile.Requests.
+// Start can be called multiple times, allowing multiple controllers to share a single source.
 func (is *Informer) Start(ctx context.Context, handler handler.EventHandler, queue workqueue.RateLimitingInterface,
 	prct ...predicate.Predicate) error {
 	// Informer should have been specified by the user.
 	if is.Informer == nil {
 		return fmt.Errorf("must specify Informer.Informer")
 	}
 
-	_, err := is.Informer.AddEventHandler(internal.NewEventHandler(ctx, queue, handler, prct).HandlerFuncs())
+	is.mu.Lock()
+	defer is.mu.Unlock()
+	if is.canceled {
+		return nil
+	}
+	if is.ctx == nil {
+		is.ctx, is.cancel = context.WithCancel(ctx)
```

Comment:
Same problem as above with Channel, only the ctx of the first Start() call is used